### DISCUSSION TITLE
Setting up custom types for Edge Attributes

### DISCUSSION THREAD

**Discussion by cloud-on-prem** [ts-graphviz/ts-graphviz#899] [Q&A] [answered]
Hi there, 

I am following your README to set up custom attributes on Edge.

I have set up a custom.d.ts file as below.

```ts
import { $keywords } from 'ts-graphviz';

declare module 'ts-graphviz' {
  export namespace EdgeAttributeKey {
    // eslint-disable-next-line @typescript-eslint/no-empty-interface
    export interface $values extends $keywords<'testDir' | 'testCaller'> { }
  }

  export namespace Attribute {
    // eslint-disable-next-line @typescript-eslint/no-empty-interface
    export interface $keys extends $keywords<'testDir' | 'testCaller'> { }

    export interface $types {
      testDir: 'pull' | 'push' | 'both';
      testCaller: string;
    }
  }
}
```

And in the code, I use it as such:

```ts
graph.attributes.edge.set(attribute.testDir, "s");
```

I do get autocompletion for `testDir` confirming the the key is correctly typed but I don't get any type errors for passing an invalid value type (`s` as opposed to `pull | push | both`)

Any thoughts?
Thanks 🙏

**Comment by kamiazya** [comment]
@cloud-on-prem 
Hi,

Thanks for the question!

I checked at hand and reproduced the problem, so there may be something unexpected.

Specifically, there seems to be a problem with the specification in set, since the type validation did not take effect even with the following.
```ts
graph.attributes.edge.set(attribute.color, 's');
````

I will convert this question to an issue and manage it.

**Reply by kamiazya** [reply]
Converting was not possible, so we responded in the form of an attachment.

We will close it as a question, as we will manage it here thereafter.

https://github.com/ts-graphviz/ts-graphviz/issues/900#issue-1762898071

**Reply by kamiazya** [reply, marked as answer]
@cloud-on-prem 

The issue has been resolved in the [v1.8.1](https://github.com/ts-graphviz/ts-graphviz/releases/tag/v1.8.1) update. We appreciate the impact of your question on enhancing the library's quality :1st_place_medal:

**Reply by cloud-on-prem** [reply]
Thanks for the quick turn-around. It works as expected. 🙏
